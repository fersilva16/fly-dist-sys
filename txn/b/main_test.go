package main

import (
	"gossip-gloomers/testutils"
	"testing"

	"github.com/gkampitakis/go-snaps/snaps"
	maelstrom "github.com/jepsen-io/maelstrom/demo/go"
	"github.com/stretchr/testify/require"
)

func Test(t *testing.T) {
	require := require.New(t)

	node = maelstrom.NewNode()

	link := testutils.NewLink(node)
	client := testutils.NewClient("c0", link)
	node1 := testutils.NewClient("n1", link)

	go main()

	err := client.InitNode("n0", []string{"n0", "n1"})

	require.NoError(err)

	txnOutput1, err := client.RPC(TxnRequest{
		MessageBody: maelstrom.MessageBody{
			Type:  "txn",
			MsgID: 2,
		},

		Txn: Txn{
			{WRITE, 1, 1},
			{READ, 2, nil},
		},
	})

	require.NoError(err)

	snaps.MatchSnapshot(t, txnOutput1)

	replicateOutput1, err := node1.Read()

	require.NoError(err)

	snaps.MatchSnapshot(t, replicateOutput1)

	err = node1.Write(ReplicateResponse{
		MessageBody: maelstrom.MessageBody{
			Type: "replicate_ok",
		},

		Keys: []int{1},
	})

	require.NoError(err)

	err = node1.Write(ReplicateRequest{
		MessageBody: maelstrom.MessageBody{
			Type: "replicate",
		},

		Clock: 1,
		Snapshot: map[int]Value{
			2: {
				Value: 2,
				TxnId: 1,
			},
		},
	})

	require.NoError(err)

	replicateOutput2, err := node1.Read()

	require.NoError(err)

	snaps.MatchSnapshot(t, replicateOutput2)

	txnOutput2, err := client.RPC(TxnRequest{
		MessageBody: maelstrom.MessageBody{
			Type:  "txn",
			MsgID: 3,
		},

		Txn: Txn{
			{READ, 1, nil},
			{WRITE, 2, 2},
		},
	})

	require.NoError(err)

	snaps.MatchSnapshot(t, txnOutput2)

	replicateOutput3, err := node1.Read()

	require.NoError(err)

	snaps.MatchSnapshot(t, replicateOutput3)

	err = node1.Write(ReplicateResponse{
		MessageBody: maelstrom.MessageBody{
			Type: "replicate_ok",
		},

		Keys: []int{2},
	})

	require.NoError(err)
}

func TestPartition(t *testing.T) {
	require := require.New(t)

	node = maelstrom.NewNode()

	link := testutils.NewLink(node)
	client := testutils.NewClient("c0", link)
	node1 := testutils.NewClient("n1", link)

	go main()

	err := client.InitNode("n0", []string{"n0", "n1"})

	require.NoError(err)

	txnOutput1, err := client.RPC(TxnRequest{
		MessageBody: maelstrom.MessageBody{
			Type:  "txn",
			MsgID: 2,
		},

		Txn: Txn{
			{WRITE, 1, 1},
			{READ, 2, nil},
		},
	})

	require.NoError(err)

	snaps.MatchSnapshot(t, txnOutput1)

	replicateOutput1, err := node1.Read()

	require.NoError(err)

	snaps.MatchSnapshot(t, replicateOutput1)

	txnOutput2, err := client.RPC(TxnRequest{
		MessageBody: maelstrom.MessageBody{
			Type:  "txn",
			MsgID: 3,
		},

		Txn: Txn{
			{READ, 1, nil},
			{WRITE, 2, 2},
		},
	})

	require.NoError(err)

	snaps.MatchSnapshot(t, txnOutput2)

	replicateOutput2, err := node1.Read()

	require.NoError(err)

	snaps.MatchSnapshot(t, replicateOutput2)

	err = node1.Write(ReplicateResponse{
		MessageBody: maelstrom.MessageBody{
			Type: "replicate_ok",
		},

		Keys: []int{1, 2},
	})

	require.NoError(err)
}
