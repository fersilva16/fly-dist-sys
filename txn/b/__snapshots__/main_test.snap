
[Test - 1]
{"src":"n0","dest":"c0","body":{"in_reply_to":2,"txn":[["w",1,1],["r",2,null]],"type":"txn_ok"}}
---

[Test - 2]
{"src":"n0","dest":"n1","body":{"type":"replicate","clock":1,"snapshot":{"1":[["w",1,1]]}}}
---

[Test - 3]
{"src":"n0","dest":"n1","body":{"in_reply_to":0,"txnIds":[2],"type":"replicate_ok"}}
---

[Test - 4]
{"src":"n0","dest":"c0","body":{"in_reply_to":3,"txn":[["r",1,2],["w",2,2]],"type":"txn_ok"}}
---

[Test - 5]
{"src":"n0","dest":"n1","body":{"type":"replicate","clock":2,"snapshot":{"3":[["w",2,2]]}}}
---

[TestPartition - 1]
{"src":"n0","dest":"c0","body":{"in_reply_to":2,"txn":[["w",1,1],["r",2,null]],"type":"txn_ok"}}
---

[TestPartition - 2]
{"src":"n0","dest":"n1","body":{"type":"replicate","clock":1,"snapshot":{"1":[["w",1,1]]}}}
---

[TestPartition - 3]
{"src":"n0","dest":"c0","body":{"in_reply_to":3,"txn":[["r",1,1],["w",2,2]],"type":"txn_ok"}}
---

[TestPartition - 4]
{"src":"n0","dest":"n1","body":{"type":"replicate","clock":2,"snapshot":{"1":[["w",1,1]],"2":[["w",2,2]]}}}
---
